### PR TITLE
File manager Refactoring

### PR DESCRIPTION
**Pull request by TonioGela**
I changed the implementation of most of the apis from `Future.fromTry{ Try{ syncBlockingStuff() } }` to `Future { syncBlockingStuff() }`.
In the case of s3 I used the s3AsyncClient and I think I managed to override the ThreadPool it uses with the ExecutionContext I'm passing to the constructor.

### CODE REVIEW

**Review by galales — APPROVED** [review]

File: file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala [hunk 1]
```diff
@@ -2,26 +2,14 @@ package it.pagopa.interop.commons.files
...
-/** Defines File manager configuration
-  */
 object StorageConfiguration {
   private lazy val config: Config =
     ConfigFactory.defaultApplication().withFallback(ConfigFactory.defaultReference()).resolve()
 
-  /** Returns the file manager name as defined in configuration file.
-    */
-  val runtimeFileManager: String = config.getString("interop-commons.storage.type")
+  lazy val maxConcurrency: Int = Option(config.getInt("interop-commons.storage.max-concurrency")).getOrElse(50)
```
[the remark this answers is not in the document]
Comment (galales) [reply file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala:11]: hello there, I'm your guilt

**Review by beetlecrunch — APPROVED** [review]

File: file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala [hunk 2]
```diff
@@ -2,26 +2,14 @@ package it.pagopa.interop.commons.files
...
-  * @param applicationId application identifier as defined on third party storage (e.g.: AWS CLIENT)
-  * @param applicationSecret application password as defined on third party storage (e.g.: AWS PASSWORD)
-  * @param endpoint third party storage location (e.g.: AWS S3 endpoint)
-  */
 case class StorageAccountInfo(applicationId: String, applicationSecret: String, endpoint: String)
 
-/** Defines File manager configuration
-  */
 object StorageConfiguration {
   private lazy val config: Config =
```
[the remark this answers is not in the document]
Comment (TonioGela) [reply file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala]: The topic is broad IMHO. On one hand, I desire to remove from every library we have this hidden lazy instantiation of `Config.load()`, to replace Typesafe Config library with PureConfig and refactor every class adding a constructor parameter that is the typed configuration. On the other hand I don't know if it's worth spending time doing it. The truth is that for this particular case I cannot foresee whether laziness is good or evil.

[the remark this answers is not in the document]
Comment (galales) [reply file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala]: one more bit: in the past we found that removing laziness from one config in the commons resulted in a service to fail to boot (I don't remember which config and which service, sorry).
I'm not against adding/removing laziness, that's just to say that we need to check that nothing will break

[the remark this answers is not in the document]
Comment (TonioGela) [reply file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala]: I remember it, but let me say that adding laziness is like raising the timeout if a service is slow to respond. That's a thing to address nevertheless, with or without the laziness keyword there. I'll add it again but we have to address these configuration reading/application boot issues once and for all (and my proposal is to use PureConfig and to pass explicitly the config in the constructors, reading it with a Reader monad like pattern)

File: file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala [hunk 3]
```diff
@@ -2,29 +2,8 @@ package it.pagopa.interop.commons.files
 
 import com.typesafe.config.{Config, ConfigFactory}
 
-/** Models configuration data for connecting to the external storage, if any.
-  *
-  * @param applicationId application identifier as defined on third party storage (e.g.: AWS CLIENT)
-  * @param applicationSecret application password as defined on third party storage (e.g.: AWS PASSWORD)
-  * @param endpoint third party storage location (e.g.: AWS S3 endpoint)
-  */
 case class StorageAccountInfo(applicationId: String, applicationSecret: String, endpoint: String)
```
Comment (galales) [inline file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala]: `applicationId` and `applicationSecret` are no more used

File: file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala [hunk 4]
```diff
@@ -2,29 +2,8 @@ package it.pagopa.interop.commons.files
...
 
-/** Defines File manager configuration
-  */
 object StorageConfiguration {
-  private lazy val config: Config =
-    ConfigFactory.defaultApplication().withFallback(ConfigFactory.defaultReference()).resolve()
-
-  /** Returns the file manager name as defined in configuration file.
-    */
-  val runtimeFileManager: String = config.getString("interop-commons.storage.type")
```
Comment (galales) [inline file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala:21]: does this mean that the storage type is no more set in the config file?

**Review by beetlecrunch — APPROVED** [review]

File: file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala (hunk 4 above)
Comment (TonioGela) [reply file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala:21]:
```scala
  def get(kind: FileManager.Kind)(blockingExecutionContext: ExecutionContextExecutor): FileManager = kind match {
    case File => new FileManagerImpl(blockingExecutionContext)
    case S3   => new S3ManagerImpl(blockingExecutionContext)
  }
```

image

File: file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala (hunk 3 above)
Comment (TonioGela) [reply file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala]: it's the whole class that is useless now in fact

Comment (TonioGela) [reply file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala]: I removed it

File: file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala (hunk 4 above)
Comment (galales) [reply file-manager/src/main/scala/it/pagopa/interop/commons/files/StorageConfiguration.scala:21]: this has nothing to do with the configs.
But ok, each service would just have it's own configuration then 👍

**Pull request merged**